include/cetl/pmr/function.hpp

```diff
@@ -0,0 +1,321 @@
+///
+namespace detail
+{
+
+// 436C9E2B-96E3-4483-9D2B-32B5147A0314
+using function_handler_typeid_t = cetl::
+    // NOLINTNEXTLINE(cppcoreguidelines-avoid-magic-numbers, readability-magic-numbers)
+    type_id_type<0x43, 0x6C, 0x9E, 0x2B, 0x96, 0xE3, 0x44, 0x83, 0x9D, 0x2B, 0x32, 0xB5, 0x14, 0x7A, 0x03, 0x14>;
+
+/// @brief Provides an abstract interface for copyable functors.
+///
+template <typename Result, typename... Args>
+class function_handler : public rtti_helper<function_handler_typeid_t>
+{
+public:
+    virtual Result operator()(Args...) = 0;
+
+};  // function_handler
+
+// DCAAADD6-BC73-4E3C-85B7-E9473641E737
+using functor_handler_typeid_t = cetl::
+    // NOLINTNEXTLINE(cppcoreguidelines-avoid-magic-numbers, readability-magic-numbers)
+    type_id_type<0xDC, 0xAA, 0xAD, 0xD6, 0xBC, 0x73, 0x4E, 0x3C, 0x85, 0xB7, 0xE9, 0x47, 0x36, 0x41, 0xE7, 0x37>;
+
+template <typename Functor, typename Result, typename... Args>
+class functor_handler final : public rtti_helper<functor_handler_typeid_t, function_handler<Result, Args...>>
+{
+public:
+    explicit functor_handler(Functor functor)
+        : functor_{functor}
+    {
+    }
+
+    Result operator()(Args... args) override
+    {
+        return functor_(args...);
+    }
+
+private:
+    Functor functor_;
+
+};  // functor_handler
+
+[[noreturn]] inline void throw_bad_function_call()
+{
+#if defined(__cpp_exceptions)
+    throw std::bad_function_call{};
+#else
+    std::terminate();
+#endif
+}
+
+template <typename Signature, std::size_t Footprint, bool IsPmr>
+class function_base;
+//
+template <typename Result, typename... Args, std::size_t Footprint>
+class function_base<Result(Args...), Footprint, false /*IsPmr*/>
+{
+public:
+    function_base() noexcept                                 = default;
+    ~function_base()                                         = default;
+    function_base& operator=(const function_base& other)     = delete;
+    function_base& operator=(function_base&& other) noexcept = delete;
+
+    function_base(const function_base& other)
+    {
+        if (other.any_handler_.has_value())
+        {
+            any_handler_ = other.any_handler_;
+            handler_ptr_ = get_if<detail::function_handler<Result, Args...>>(&any_handler_);
+
+            CETL_DEBUG_ASSERT(any_handler_.has_value(), "");
+            CETL_DEBUG_ASSERT(nullptr != handler_ptr_, "");
+        }
+    }
+
+    function_base(function_base&& other) noexcept
+    {
+        if (other.any_handler_.has_value())
+        {
+            any_handler_       = std::move(other.any_handler_);
+            handler_ptr_       = get_if<detail::function_handler<Result, Args...>>(&any_handler_);
+            other.handler_ptr_ = nullptr;
+
+            CETL_DEBUG_ASSERT(any_handler_.has_value(), "");
+            CETL_DEBUG_ASSERT(nullptr != handler_ptr_, "");
+        }
+    }
+
+    // TODO: Add Callable constraint;
```

Comment:
These "Add Callable constraint" todos will be addressed separately. I believe it can be properly done by adding `cetl::pf17::invoke` support (like `cetl::is_invokable<...>` stuff).